### ISSUE TITLE
如何使用多GPU训练？

### ISSUE THREAD

**Issue by SasAILab** [anosorae/IRRA#40]
如何使用多GPU训练？服务器有四张卡（0，1，2，3）.然后我想使用卡1和卡2.怎么办？求大神告知，我看源码里面好像对多GPU训练的部分写的有些问题，有人遇到过这个问题吗

**Comment by SasAILab** [comment]
已经解决了,以前用的都是DataParallel，而本代码里面用的是DistributedDataParallel来做并行训练。
准备过几天好好把并行训练的知识点学一学！我服务器上有4张卡(每张49G),然后我用了两张卡跑,batch设置为188
下面是我使用的两个命令，第一个是放在后台离线跑，指定输出文件。第二个是在线跑，在终端打印结果
nohup sh -c "CUDA_VISIBLE_DEVICES=1,2 python -m torch.distributed.launch --nproc_per_node=2 multi_gpu_train.py --name='irra' --img_aug --MLM --dataset_name='CUHK-PEDES' --loss_names='sdm+mlm+id' --root_dir='./data' --num_epoch=100 --batch_size=188" > run-IRRA-epoch100.log 2>&1 &

CUDA_VISIBLE_DEVICES=1,2 python -m torch.distributed.launch --nproc_per_node=2 multi_gpu_train.py --name='irra' --img_aug --MLM --dataset_name='CUHK-PEDES' --loss_names='sdm+mlm+id' --root_dir='./data' --num_epoch=100 --batch_size=188

**Comment by koalayyy** [comment]
请问使用多gpu时代码里做了哪些修改呢，现在跑多gpu也遇到了一些问题，感谢！

**Comment by SasAILab** [comment]
你好，代码没做任何修改，我是加了一个sh脚本，sh脚本命名为run.sh，其内容如下：
#!/home/202312150002/anaconda3/envs/lzm/bin/python3.8
export CUDA_VISIBLE_DEVICES=0,1,2 

python -m torch.distributed.run --nproc_per_node=3 \
train.py \
    --name style4_global_itc2 \
    --img_aug \
    --batch_size 128 \
    --MLM \
    --pretrain_choice 'ViT-B/16' \
    --loss_names 'sdm+mlm+id' \
    --num_epoch 100

**Comment by lizimo-n** [comment]
上面的脚本内容，后面都有一个反斜杠哈，github里面没有显示出来好像

**Comment by SasAILab** [comment]
上面的脚本内容，从python-m那一行开始后面都有一个反斜杠哈。最后一行不用，刚才我发现github里面没有显示出来这个反斜杠，不要误导您

**Comment by koalayyy** [comment]
> 上面的脚本内容，从python-m那一行开始后面都有一个反斜杠哈。最后一行不用，刚才我发现github里面没有显示出来这个反斜杠，不要误导您

好的，非常感谢！

**Comment by sunzhens** [comment]
> 已经解决了,以前用的都是DataParallel，而本代码里面用的是DistributedDataParallel来做并行训练。 准备过几天好好把并行训练的知识点学一学！我服务器上有4张卡(每张49G),然后我用了两张卡跑,batch设置为188 下面是我使用的两个命令，第一个是放在后台离线跑，指定输出文件。第二个是在线跑，在终端打印结果 nohup sh -c "CUDA_VISIBLE_DEVICES=1,2 python -m torch.distributed.launch --nproc_per_node=2 multi_gpu_train.py --name='irra' --img_aug --MLM --dataset_name='CUHK-PEDES' --loss_names='sdm+mlm+id' --root_dir='./data' --num_epoch=100 --batch_size=188" > run-IRRA-epoch100.log 2>&1 &
> 
> CUDA_VISIBLE_DEVICES=1,2 python -m torch.distributed.launch --nproc_per_node=2 multi_gpu_train.py --name='irra' --img_aug --MLM --dataset_name='CUHK-PEDES' --loss_names='sdm+mlm+id' --root_dir='./data' --num_epoch=100 --batch_size=188


你好，我尝试用下面这个命令用多GPU复现
CUDA_VISIBLE_DEVICES=0,1,2 python -m torch.distributed.run --nproc_per_node=3 train.py --name style4_global_itc2 --img_aug --batch_size 128 --MLM --pretrain_choice 'ViT-B/16' --loss_names 'sdm+mlm+id' --root_dir '/home/xmu/datasets_lwh/'  --num_epoch 100 --dataset_name 'CUHK-PEDES'
但是出现了下面的问题，请问这怎么解决啊?谢谢
![image](https://github.com/anosorae/IRRA/assets/90076810/7c887bc9-d34c-4b29-8fe5-e3feffe08df2)

**Comment by yiranxie233** [comment]
我是用了两张12G的3060跑，用了楼主的指令训练，发现了报错， 后来尝试使用指令,可以在我的机子上面运行
~~~python
CUDA_VISIBLE_DEVICES=0,1 python -m torch.distributed.launch --nproc_per_node=2 train.py --MLM --name='irra'
~~~
需要对原来的utils/options.py的第一个配置参数进行修改：
~~~python
# 原
# parser.add_argument("--local_rank", default=0, type=int)
# 改
parser.add_argument("--local-rank", default=0, type=int)
~~~

**Comment by ForeUP** [comment]
> 你好，代码没做任何修改，我是加了一个sh脚本，sh脚本命名为run.sh，其内容如下： #!/home/202312150002/anaconda3/envs/lzm/bin/python3.8 export CUDA_VISIBLE_DEVICES=0,1,2
> 
> python -m torch.distributed.run --nproc_per_node=3 train.py --name style4_global_itc2 --img_aug --batch_size 128 --MLM --pretrain_choice 'ViT-B/16' --loss_names 'sdm+mlm+id' --num_epoch 100

作者的代码里并没有实现ddp训练的sampler，应该需要自己补充一下，如下面代码所示：
https://github.com/anosorae/IRRA/blob/c698f851e5c866cd79031320825844e547249690/datasets/build.py#L116-L118

我尝试改了下，但是具体是不是存在错误需要等我训练完验证一下：

```python
            if args.distributed:
                # TODO valid distributed condition
                logger.info('using ddp random sampler')
                logger.info('DISTRIBUTED TRAIN START')
                mini_batch_size = args.batch_size // get_world_size()
                # 初始化DistributedSampler
                train_sampler = DistributedSampler(train_set, shuffle=True)
                train_loader = DataLoader(train_set,
                                          batch_size=mini_batch_size,  # 改为mini_batch
                                          sampler=train_sampler,  # 使用DistributedSampler
                                          num_workers=num_workers,
                                          collate_fn=collate)
```

**Comment by ThomaswellY** [comment]
请问你们多卡能跑出paper中的结果吗？我的复现结果差的很多，我只能认为是多卡DDP需要更高的学习率了（代码中的lr=1e-5，但是论文说用单卡跑的）

**Comment by ForeUP** [comment]
> 请问你们多卡能跑出paper中的结果吗？我的复现结果差的很多，我只能认为是多卡DDP需要更高的学习率了（代码中的lr=1e-5，但是论文说用单卡跑的）

我单卡bath-size设成32，和论文结果接近，平均差1个点左右，双卡我也没改出来，双卡很难或几乎不可能达到单卡的指标
CUHK：
```
2024-08-01 21:20:53,674 IRRA.train INFO: Validation Results - Epoch: 54
2024-08-01 21:21:07,212 IRRA.eval INFO: 
+------+--------+--------+--------+--------+--------+
| task |   R1   |   R5   |  R10   |  mAP   |  mINP  |
+------+--------+--------+--------+--------+--------+
| t2i  | 72.807 | 89.263 | 93.583 | 66.351 | 51.366 |
+------+--------+--------+--------+--------+--------+
```
ICFG：
```
2024-08-03 05:23:54,249 IRRA.train INFO: Validation Results - Epoch: 60
2024-08-03 05:25:36,209 IRRA.eval INFO: 
+------+--------+--------+--------+--------+-------+
| task |   R1   |   R5   |  R10   |  mAP   |  mINP |
+------+--------+--------+--------+--------+-------+
| t2i  | 62.253 | 79.373 | 85.046 | 38.221 | 8.421 |
+------+--------+--------+--------+--------+-------+
```
RSTP：
```
2024-08-03 02:02:55,567 IRRA.train INFO: Validation Results - Epoch: 47
2024-08-03 02:02:59,976 IRRA.eval INFO: 
+------+--------+--------+--------+--------+--------+
| task |   R1   |   R5   |  R10   |  mAP   |  mINP  |
+------+--------+--------+--------+--------+--------+
| t2i  | 59.050 | 79.550 | 86.450 | 47.630 | 27.144 |
+------+--------+--------+--------+--------+--------+
```

**Comment by lyx-JuneSnow** [comment]
> > 你好，代码没做任何修改，我是加了一个sh脚本，sh脚本命名为run.sh，其内容如下： #!/home/202312150002/anaconda3/envs/lzm/bin/python3.8 export CUDA_VISIBLE_DEVICES=0,1,2
> > python -m torch.distributed.run --nproc_per_node=3 train.py --name style4_global_itc2 --img_aug --batch_size 128 --MLM --pretrain_choice 'ViT-B/16' --loss_names 'sdm+mlm+id' --num_epoch 100
> 
> 作者的代码里并没有实现ddp训练的sampler，应该需要自己补充一下，如下面代码所示：
> 
> [IRRA/datasets/build.py](https://github.com/anosorae/IRRA/blob/c698f851e5c866cd79031320825844e547249690/datasets/build.py#L116-L118)
> 
> Lines 116 to 118 in [c698f85](/anosorae/IRRA/commit/c698f851e5c866cd79031320825844e547249690)
> 
>  elif args.sampler == 'random': 
>      # TODO add distributed condition 
>      logger.info('using random sampler') 
> 我尝试改了下，但是具体是不是存在错误需要等我训练完验证一下：
> 
>             if args.distributed:
>                 # TODO valid distributed condition
>                 logger.info('using ddp random sampler')
>                 logger.info('DISTRIBUTED TRAIN START')
>                 mini_batch_size = args.batch_size // get_world_size()
>                 # 初始化DistributedSampler
>                 train_sampler = DistributedSampler(train_set, shuffle=True)
>                 train_loader = DataLoader(train_set,
>                                           batch_size=mini_batch_size,  # 改为mini_batch
>                                           sampler=train_sampler,  # 使用DistributedSampler
>                                           num_workers=num_workers,
>                                           collate_fn=collate)

您好，我尝试了这样修改代码，遇到报错为RuntimeError: DDP expects same model across all ranks, but Rank 0 has 367 params, while rank 1 has inconsistent 0 params.  但没有发现问题所在，请问大佬当时跑的时候有问题吗？我是把您提供的代码加在了datasets/build.py中并注释了一部分您提到的源代码

**Comment by AuKeeWa** [comment]
> > > 你好，代码没做任何修改，我是加了一个sh脚本，sh脚本命名为run.sh，其内容如下： #!/home/202312150002/anaconda3/envs/lzm/bin/python3.8 export CUDA_VISIBLE_DEVICES=0,1,2
> > > python -m torch.distributed.run --nproc_per_node=3 train.py --name style4_global_itc2 --img_aug --batch_size 128 --MLM --pretrain_choice 'ViT-B/16' --loss_names 'sdm+mlm+id' --num_epoch 100
> > 
> > 
> > 作者的代码里并没有实现ddp训练的sampler，应该需要自己补充一下，如下面代码所示：
> > [IRRA/datasets/build.py](https://github.com/anosorae/IRRA/blob/c698f851e5c866cd79031320825844e547249690/datasets/build.py#L116-L118)
> > Lines 116 to 118 in [c698f85](/anosorae/IRRA/commit/c698f851e5c866cd79031320825844e547249690)
> > elif args.sampler == 'random':
> > # TODO add distributed condition
> > logger.info('using random sampler')
> > 我尝试改了下，但是具体是不是存在错误需要等我训练完验证一下：
> > ```
> >         if args.distributed:
> >             # TODO valid distributed condition
> >             logger.info('using ddp random sampler')
> >             logger.info('DISTRIBUTED TRAIN START')
> >             mini_batch_size = args.batch_size // get_world_size()
> >             # 初始化DistributedSampler
> >             train_sampler = DistributedSampler(train_set, shuffle=True)
> >             train_loader = DataLoader(train_set,
> >                                       batch_size=mini_batch_size,  # 改为mini_batch
> >                                       sampler=train_sampler,  # 使用DistributedSampler
> >                                       num_workers=num_workers,
> >                                       collate_fn=collate)
> > ```
> 
> 您好，我尝试了这样修改代码，遇到报错为RuntimeError: DDP expects same model across all ranks, but Rank 0 has 367 params, while rank 1 has inconsistent 0 params. 但没有发现问题所在，请问大佬当时跑的时候有问题吗？我是把您提供的代码加在了datasets/build.py中并注释了一部分您提到的源代码

您好，我也遇到了同样的问题，请问您现在解决了吗？

**Comment by AuKeeWa** [comment]
> 请问你们多卡能跑出paper中的结果吗？我的复现结果差的很多，我只能认为是多卡DDP需要更高的学习率了（代码中的lr=1e-5，但是论文说用单卡跑的）

您好，请问您的多卡训练是否有对代码进行修改呢？非常感谢！
